sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ExecutionStateSampler.java

```diff
@@ -350,21 +400,58 @@ private void takeSample(long currentTimeMillis, long millisSinceLastSample) {
         currentExecutionState.takeSample(millisSinceLastSample);
       }
 
+      Thread thread = trackedThread.get();
```

Comment:
keep this inside the ifs if the lull time is great enough, we want the sampling to be as lightweight as possible in the common case and should avoid this atomic overhead if unneeded